sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -762,7 +716,20 @@ class RocksDB(
 
   def doMaintenance(): Unit = {
     if (enableChangelogCheckpointing) {
-      uploadSnapshot()
+
+      var mostRecentSnapshot: Option[RocksDBSnapshot] = None
+      var snapshot = snapshots.poll()
+
+      while (snapshot != null) {
```

Comment:
Could we just add entries to an array, keep the last one and close the earlier ones ?

Reply:
I don't think that is any better. Also don't think we should be allocating a new Array just for this. Below is a snippet of how that would look like and doesn't look any better. I've added comment, so hopefully that helps

```
val snapshots = new ArrayBuffer[RocksDBSnapshot]()
var snapshot = snapshotsToUploadQueue.poll()

while (snapshot != null) {
  logDebug(s"RocksDB Maintenance - polled snapshot ${snapshot.version}")
  snapshots += snapshot
  snapshot = snapshotsToUploadQueue.poll()
}

val mostRecentSnapshot = snapshots.lastOption
snapshots.init.foreach(_.close()) // close all snapshots except the last one
```

Reply:
yeah, even if new one is added by task thread, we will upload the new one.